jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java

```diff
@@ -215,13 +218,27 @@ public void clearProcessInstances() {
 
     public void clearProcessInstancesState() {
         try {
+            // this happens during rollback. If timers is not transactional (ejb marked as default then the timer has not being registered
+            // therefore there is no gargabe to take care. The cleanup is happening during TransactionAwareSchedulerServiceInterceptor rollback status
+            // if there is garbage in the timer manager as we have there the timer instance
+            // Otherwise we let this happen
+
             // at this point only timers are considered as state that needs to be cleared
             TimerManager timerManager = ((InternalProcessRuntime)kruntime.getProcessRuntime()).getTimerManager();
-            
+            TimerService timerService = timerManager.getTimerService();
+            if(timerService instanceof RegisteredTimerServiceDelegate) {
+                timerService = ((RegisteredTimerServiceDelegate) timerService).getTimerService();
+            }
+            if(timerService instanceof GlobalTimerService) {
+                // if the timer instance is transactional we don't need to clean up the timers as the rollback part of the transaction awareness
+                // is happening
+                if(!((GlobalTimerService) timerService).isTransactional()) {
```

Comment:
@elguardian shouldn't this method return if the timerService *is* actually transactional? If yes, the ! negation should be removed here imo.

Reply:
well... in spite it seems. the answer is no. When is not transactinal the clean up is done in another place (interceptor).
This is a bit strange. This clean up was done 9 years ago while more modern parts (regarding interceptors were done few years later. Also ejb has a flag that is totally out of place (true/false if they can be executed in tx when ejb are always executed within tx. 
Tbh honest this clean up should be removed but at this point it is not really safe.
so this is the best I can do. move the clean up when it is not transactional to the interceptor and remove that situation from here.
This saves the rest of the cases like quartz in spring boot and jdk thread pools

Reply:
I tested this again and it does not work for me. Running in kie-server, isTransactional is true, so the cleanup code is still called.